### PR TITLE
PA-4 - Create Api service [v0.5.0]

### LINKED ISSUE #8: PA-4 - Create api service

### PR DESCRIPTION
**Pull request by iamdavidpetric**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
   "dependencies": {
+    "@heroicons/react": "2.0.16",
     "@testing-library/jest-dom": "5.16.5",
     "@testing-library/react": "13.4.0",
     "@testing-library/user-event": "13.5.0",
+    "axios": "1.3.4",
     "env-cmd": "10.1.0",
     "react": "18.2.0",
     "react-dom": "18.2.0",
+    "react-icons": "4.7.1",
```
Comment (iamalextodoran) [inline package.json]: Nono

File: src/logic/api/index.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import axios from 'axios';
+
+const defaultConfig = {
+  baseUrl: `${process.env.REACT_APP_API_URL}${process.env.REACT_APP_API_PATH_PREFIX}`,
+  timeout: 10_000,
```
Comment (iamalextodoran) [inline src/logic/api/index.js]: Set the timeout in env files (without the `_`)

File: src/logic/api/index.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+const api = axios.create(defaultConfig);
+
+const ownInterseptors = {
+  handleRequestError: error => {
+    return Promise.reject(error);
+  },
+  handleResponseError: error => {
+    return Promise.reject(error);
+  },
+  handleResponse: response => {
```
Comment (iamalextodoran) [inline src/logic/api/index.js]:
```suggestion
  handleResponse: config => {
```

File: src/logic/api/index.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  timeout: 10_000,
+  headers: {
+    Accept: 'application/json',
+    'Content-Type': 'application/json',
+  },
+};
+
+const api = axios.create(defaultConfig);
+
+const ownInterseptors = {
```
Comment (iamalextodoran) [inline src/logic/api/index.js]: typo

File: src/view/navigator/index.js [hunk 5]
```diff
@@ -18,13 +18,15 @@ const renderRoute = (isLoggedIn, type, route) => {
...
 
 const Navigator = ({ isLoggedIn = false }) => (
-  <BrowserRouter>
-    <Routes>
-      {publicRoutes.map(renderRoute.bind(null, isLoggedIn, 'public'))}
-      {privateRoutes.map(renderRoute.bind(null, isLoggedIn, 'private'))}
-      {neutralRoutes.map(renderRoute.bind(null, isLoggedIn, 'neutral'))}
-    </Routes>
-  </BrowserRouter>
+  <div className='min-h-screen h-screen w-screen'>
```
Comment (iamalextodoran) [inline src/view/navigator/index.js]: This should go into application. Revert this file.

File: src/view/pages/public/Home/index.js [hunk 6]
```diff
@@ -1,3 +1,45 @@
...
+
+import Api from '../../../../logic/api';
+
+const errors = (setError, err) => setError(err);
+const loading = setIsLoading => setIsLoading(false);
+const response = (setPost, res) => setPost(res?.data);
+
+const Home = () => {
+  const [post, setPost] = useState([]);
+  const [error, setError] = useState([]);
```
Comment (iamalextodoran) [inline src/view/pages/public/Home/index.js]: Wrong defaults

File: src/logic/api/index.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+api.interceptors.response.use(ownInterseptors.handleResponse, ownInterseptors.handleResponseError);
+
+const Api = {
+  API_CALL: 'API_CALL',
+
+  catchError: error => {
+    return error.response.data;
+  },
+
+  delete: path => {
```
Comment (iamalextodoran) [inline src/logic/api/index.js]: For each crud method add a new parameter called config and initialise it with `defaultConfig`, then, where you call the method add config as the last argument.

**Notice the difference between `argument` and `parameter`** -- look it up on Google

File: src/view/pages/public/Home/index.js [hunk 8]
```diff
@@ -1,3 +1,45 @@
...
+  const [counter, setCounter] = useState(1);
+  const [isLoading, setIsLoading] = useState(false);
+
+  useEffect(() => {
+    setIsLoading(true);
+    Api.get(`https://jsonplaceholder.typicode.com/posts/${counter}`)
+      .then(response.bind(null, setPost))
+      .catch(errors.bind(null, setError))
+      .finally(loading.bind(null, setIsLoading));
```
Comment (iamalextodoran) [inline src/view/pages/public/Home/index.js]: You don't need to create a new function that uses setLoading, just bind setLoading properly

**Review by iamalextodoran — CHANGES_REQUESTED** [review]

File: src/view/pages/public/Home/index.js [hunk 9]
```diff
@@ -1,3 +1,45 @@
...
+  }, [counter]);
+
+  return (
+    <div className='flex flex-col h-full'>
+      <div>
+        {JSON.stringify(post)}
+        {JSON.stringify(error)}
+      </div>
+      <div className='flex flex-col justify-center items-center '>
```
Comment (iamdavidpetric) [inline src/view/pages/public/Home/index.js]:
```suggestion
      <div className='flex flex-col justify-center items-center'>
```
https://tailwindcss.com/blog/automatic-class-sorting-with-prettier

Comment (iamalextodoran) [reply src/view/pages/public/Home/index.js]: Yes!

File: .env.development [hunk 10]
```diff
@@ -1,4 +1,5 @@
 PORT=8080
 REACT_APP_API_PATH_PREFIX='/v1'
 REACT_APP_API_URL=http://localhost:3000
-REACT_APP_ENVIRONMENT=development
\ No newline at end of file
+REACT_APP_ENVIRONMENT=development
+REACT_APP_TIMEOUT=10000
```
Comment (iamalextodoran) [inline .env.development:5]: What about the other envs, don't they also need a timeout?

File: src/logic/api/index.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    const request = api.delete(path, config);
+    request.catch(Api.catchError);
+    return request;
+  },
+
+  get: (path, config = defaultConfig) => {
+    const request = api.get(path, config);
+    request.catch(Api.catchError);
+    return request;
+  },
```
Comment (iamalextodoran) [inline src/logic/api/index.js]:
```suggestion
  get: (path, params = {}, config = defaultConfig) => {
    const request = api.get(path, { ...config, params });
    request.catch(Api.catchError);
    return request;
  },
```
Neat trick to have query params right into the get method

File: src/view/application/index.jsx [hunk 12]
```diff
@@ -1,9 +1,9 @@
 import Navigator from '../navigator';
 
 const Application = () => (
-  <>
+  <div className='h-screen min-h-screen w-screen'>
```
Comment (iamalextodoran) [inline src/view/application/index.jsx]: Revert this file and apply the h-screen class directly for the body html tag into your style.css file

**Review by iamalextodoran — CHANGES_REQUESTED** [review]

**Review by iamalextodoran — APPROVED** [review]
🌮

**Pull request merged**